File: biconnected_components/typescript/adjacency_list.ts
export interface Graph {
    vertexCount: number;
    directed: boolean;
    edges: number[][];
}

export class AdjacencyList {
    a: number[][];
    constructor(graph: Graph) {
        this.a = constructAdjacencyList(graph);
    }
}

function constructAdjacencyList(graph: Graph): number[][] {
    // new Array(vertexCount).fill([]) does not work because it reuses the same array instance
    // for every element.
    const a = Array.from({length: graph.vertexCount}, () => []);

    for (const edge of graph.edges) {
        insertEdge(a, edge[0], edge[1], graph.directed);
    }

    return a;
}

function insertEdge(a: number[][], x: number, y: number, directed: boolean): void {
    a[x].push(y);
    if (!directed) {
        insertEdge(a, y, x, true);
    }
}

File: maximal_clique/go/adjacency_list.go
package maximal_clique

type Graph struct {
	VertexCount int
	Edges       []Edge
}

type Edge struct {
	X int
	Y int
}

type AdjacencyList [][]int

func NewAdjacencyList(g Graph) AdjacencyList {
	a := make(AdjacencyList, g.VertexCount)
	for _, edge := range g.Edges {
		a.insertEdge(edge)
	}
	return a
}

func (a AdjacencyList) insertEdge(edge Edge) {
	a[edge.X] = append(a[edge.X], edge.Y)
	a[edge.Y] = append(a[edge.Y], edge.X)
}
